### ISSUE TITLE
☂ Allow Twisted programs to behave correctly with respect to changes in the system clock

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2424]
|<img alt="rfg007's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| rfg007 reported|
|-|-|
|Trac ID|trac#2424|
|Type|defect|
|Created|2007-01-26 15:52:14Z|

Twisted currently assumes that the wall clock and system clock times are perfectly synchronized, and provides no facility for compensating when that stops being true.

This is a problem because the mechanism for "when to wake up" for `callLater` is decided by passing a `timeout` parameter to the operating system multiplexing API - `select()` (or similar).  This API generally considers "time" to be "the monotonic clock".

This can manifest in a few ways:

  - There's no way to schedule a repeated call - using `LoopingCall` or otherwise - which reliably runs in the face of the system clock being moved backward.  If you are trying to schedule a call every 10 milliseconds, and the system clock is moved backward by a day, the timestamp inside the `callLater` will indicate that it should not run again until 1 day plus 10 milliseconds have passed.
  - If you set a 60 minute timeout with `callLater`, wait 30 minutes, and then put your system to sleep for 120 minutes, when it wakes up it will wait an additional 30 minutes to trigger the call, _unless_ there's some other activity which wakes up the reactor, in which case it will fix itself and get called immediately.
  - If NTP introduces a large slew, the frequency of LoopingCall invocations will slow down or speed up accordingly.

To fix this, we will need to:

 1. introduce 2 new APIs: `callAfter` and `callAt`, which use the monotonic and wall clocks respectively
 2. try to monitor for notifications about "wake from sleep" and "system clock changed" events, on Windows, Mac, Linux, and FreeBSD, to the extent that that's possible.
 3. expose a new API, `callWhenTimeChanged` or something like that, to detect deltas between monotonic and wall clocks and allow applications to make specific corrections for them.

A ticket should be filed for each of these pieces of work.

Attachments:

* [2424-02-use-monotonic-clock.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/963/96327d5521a52aff95ad8066420f715fdd0d2561/0dc9c496a96bfb849396bdb6eda3dc2cc551ea33.patch) (3789 bytes) - added by ivank on 2010-05-30 22:45:36Z - 
* [linux_and_osx.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/963/96327d5521a52aff95ad8066420f715fdd0d2561/14787c6c99d22caf8cb0c78c786eccbb9215ed70.py) (1855 bytes) - added by glyph on 2012-03-25 08:08:19Z - monotonic clock for linux & OS X

<details><summary>Searchable metadata</summary>

```
trac-id__2424 2424
type__defect defect
reporter__rfg007 rfg007
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__ 
time__1169826734000000 1169826734000000
changetime__1610489299890399 1610489299890399
version__None None
owner__ 
cc__spiv cc__ivank cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

That loop could definitely be improved.  It's really, really hard to fix this completely, though.  The problem is that callLater's behavior is affected by changes in the system time.  If ```callLater(2, f)``` happens, and then one second later the system clock is set back five seconds, six more seconds must pass before ```f``` will be called.  There may also be a problem with the line you pointed out, but even if that is fixed, this ```callLater``` behavior will remain.

Fixing this may require using something like the POSIX monotonic clock.  However, accessing this from Python is difficult, integrating it with ```select(2)``` is non-obvious, and I'm not sure what equivalents exist on non-POSIX platforms.

**Comment by twisted-trac** [comment]
|<img alt="rfg007's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|rfg007 commented|
|-|-|

Thanks Glyph,

I was looking in the code what you say, and I understand the problem. I saw the implementation of ```callLater``` and ```DelayedCall```. Maybe if I introduce a simplification to my problem, you can guide me in a workaround.

Supose that the system time don't change automatically and, my application knows when to change the system time, and can change it. That is, it knows the actual system time and the required one. 

Q: can be all the pending tasks re-scheduled according to the remaining time to complete them, after the app change the time?

PD: for simplicity, I omit in the OP that I use ```taskLoopingCall``` and ```tksuport``` in the reactor, and tk use its internal timers ```.after(xx_ms, callback)``` for other updates in the gui.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

[#3899](https://github.com/twisted/twisted/issues/3899) has been closed as a duplicate of that one. I'm wondering if [#6718](https://github.com/twisted/twisted/issues/6718) wouldn't solve this as well.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|ivank commented|
|-|-|

If Twisted eventually implements a monotonic clock to fix this, it would be a breaking change, because it would cause `DelayedCall.getTime` to sometimes return the monotonic clock time. Some programs assume `dc.getTime()` returns something close to the wall-clock time. This includes Twisted (see [#4198](https://github.com/twisted/twisted/issues/4198)) and Foolscap (as of `r608:836b3aececec`), in foolscap/reconnector.py. I assume other third-party programs would be affected as well. As glyph mentioned in [#4198](https://github.com/twisted/twisted/issues/4198), the correct thing for `DelayedCall` users is (probably) to use `dc.seconds()` instead of `time.time()` when doing arithmetic with `dc.getTime()` results.

One solution is to add some flag to enable the monotonic clock, but I don't think most people will find the flag before they suffer a `DelayedCall` issue when their clock jumps.

As a side note, `libev` has some interesting code that detects clock jumps and adjusts delayed calls, even without access to a monotonic clock.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Making this a reactor option has all sorts of problems:

  * How do you specify it (most programs don't accept any arguments for reactor configuration; twistd and trial stand out in particular here; there's also a plugin API for reactors which doesn't accept any configuration)?
  * What if application code _wants_ to follow the system clock?
  * What if two different pieces of application code in one process want different things?

Rather, introducing a new API with the new desired behavior makes more sense, I think.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_7):
> Making this a reactor option has all sorts of problems:
> 
>   * How do you specify it (most programs don't accept any arguments for reactor configuration; twistd and trial stand out in particular here; there's also a plugin API for reactors which doesn't accept any configuration)?
>   * What if application code _wants_ to follow the system clock?
>   * What if two different pieces of application code in one process want different things?
> 
> Rather, introducing a new API with the new desired behavior makes more sense, I think.

+1.

Monotonic clocks are super useful, but they are something really distinct from `callLater`.  Granted, many uses of `callLater` _would_ probably prefer a monotonic clock, but we should leave it up to them to decide that explicitly.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|ivank commented|
|-|-|

I will note that nearly all existing code is using callLater for timeouts and equally-spaced actions (which should be scheduled by a monotonic clock). My perspective is that not using the monotonic clock by default is a bug, but unfortunately that is not backwards-compatible.

1. I guess the two options are: (1) add an argument to every program that eventually calls reactor.run, or (2) check os.environ['TWISTED_ALLOW_MONOTONIC']

2. If any application code wants to use callLater to follow the system clock, the developer must not enable "reactor with monotonic clock". In the future there could be some new API to follow the system clock that works even when "reactor with monotonic clock" is enabled.

3. Newly-written code should support both normal "reactor" and "reactor with monotonic clock". If any applications a developer wants to run in their process do not work with "reactor with monotonic clock", the developer must not enable the option.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|ivank commented|
|-|-|

I am attaching a patch that makes the reactor use a monotonic clock when available, with no configuration options. It is useful for checking which applications break if the reactor uses a monotonic clock. The patch also includes a fix for [#4198](https://github.com/twisted/twisted/issues/4198).

The patch requires Monoclock from http://ludios.net/git/ , which probably only works on Linux.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

I propose:

  1. Use ctypes to implement this, so it works on pypy.
  1. Monotonic time is always on.
  1. `DelayedCall.getTime` makes a best effort to figure out a wallclock time... but is deprecated.
  1. The replacement is a method called `getRemainingTime` that returns a relative time, and it's documented to be relative to `reactor.seconds` only.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_14"></a>|@zooko commented|
|-|-|

comment:13 all sounds good to me!

Here is something which might be useful to an implementor of this.

The new C++ standard defines something called ```steady_clock``` which is even better than a monotonic clock. A monotonic clock is one that can never be adjusted backward to an earlier setting. A steady clock is one that can never be adjusted.

http://en.cppreference.com/w/cpp/chrono

The reason this might be useful to an implementor of this ticket is that you might be able to crib some bits from an implementation, such as that of the Boost library.

Here is the header file:

http://svn.boost.org/svn/boost/trunk/boost/chrono/config.hpp

This shows that Boost implements ```steady_clock``` on Windows, Mac OS X, and "sun", and on POSIX if any of:
* ```CLOCK_REALTIME``` and ```CLOCK_MONOTONIC``` are defined, or
* ```_POSIX_THREAD_CPUTIME```, or
* ```CLOCK_THREAD_CPUTIME_ID```

The implementations are in ```boost/chrono/detail/inlined```.

Here are the cores of the implementations. (I've trimmed off some macros, namespaces, error-checking, and optimizations and haven't tested these. Best use the original Boost code for any actual use -- the following is just for illustration.):

* [the windows implementation](http://svn.boost.org/svn/boost/trunk/boost/chrono/detail/inlined/win/chrono.hpp) 
```
  double get_nanosecs_per_tic()
  {
      boost::detail::win32::LARGE_INTEGER_ freq;
      boost::detail::win32::QueryPerformanceFrequency( &freq );
      return double(1000000000.0L / freq.QuadPart);
  }

  steady_clock::time_point steady_clock::now()
  {
    static double nanosecs_per_tic = get_nanosecs_per_tic();

    boost::detail::win32::LARGE_INTEGER_ pcount;
    boost::detail::win32::QueryPerformanceCounter( &pcount );

    return steady_clock::time_point(steady_clock::duration(
      static_cast<steady_clock::rep>((nanosecs_per_tic) * pcount.QuadPart)));
  }
```
* [the mac implementation](http://svn.boost.org/svn/boost/trunk/boost/chrono/detail/inlined/mac/chrono.hpp)
```
double
compute_steady_factor()
{
    mach_timebase_info_data_t MachInfo;
    mach_timebase_info(&MachInfo);
    return static_cast<double>(MachInfo.numer) / MachInfo.denom;
}

steady_clock::rep
steady_full()
{
    static const double factor = chrono_detail::compute_steady_factor();
    return static_cast<steady_clock::rep>(mach_absolute_time() * factor);
}
```
* [the posix implementation](http://svn.boost.org/svn/boost/trunk/boost/chrono/detail/inlined/posix/chrono.hpp)
```
  steady_clock::time_point steady_clock::now()
  {
    timespec ts;
    ::clock_gettime( CLOCK_MONOTONIC, &ts );
    return time_point(duration(
      static_cast<steady_clock::rep>( ts.tv_sec ) * 1000000000 + ts.tv_nsec));
  }
```

(Hm, I don't see the implementation on posix when ```CLOCK_MONOTONIC``` *isn't* available...)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

> Use ctypes to implement this, so it works on pypy.

Using ctypes may just guarantee that it doesn't reliably work anywhere, not that it reliably works on CPython and PyPy. ;)  For example, there's no way to find the value of `CLOCK_MONOTONIC` using ctypes.  This is not to say that I think we should have an implementation of this in C, just to point out that ctypes has serious drawbacks as well.

> Monotonic time is always on. 

I'm not sure what this means.  Perhaps it means that we change the behavior of `reactor.callLater` to mean "call this after N seconds", rather than "call this at wallclock time now plus N seconds"?  Why would be do that, rather than introduce a second scheduling API?  Phrase your answer without contracting CompatibilityPolicy.

> DelayedCall.getTime makes a best effort to figure out a wallclock time... but is deprecated. 

This could be okay, perhaps.  On the other hand, why deprecate wall clock time support?  Can it be implemented without help from the reactor?

> The replacement is a method called getRemainingTime that returns a relative time, and it's documented to be relative to reactor.seconds only. 

Yay, new APIs.  That's more like it.  This implies that `reactor.seconds` changes its meaning as well, though.  Perhaps it should be left alone and a new monotonic time source should be introduced instead?

A separate point which I don't see addressed anywhere on this ticket, yet, is how `CLOCK_MONOTONIC` interacts with the timeout parameter to `select`, `epoll_wait`, `kevent`, etc.  Perhaps this is obvious to everyone else and there's no special consideration required in the implementation; if so, it would be nice for someone to say so. :)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst commented|
|-|-|

I am unhappy with a new API because it means everyone has to change their existing code. OTOH a pluggable time policy (or changed default) might in theory break existing code, yes.

As far as `epoll_wait` and friends, it looks like one possibility is to use a `timerfd` with CLOCK_MONOTONIC (though that's Linux only). Moreover libevent has monotonic time support with epoll et al so probably we can look at their implementation as an example. And quite possibly their support is cross-platform so educational in general.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_17"></a>|@zooko commented|
|-|-|

Replying to [exarkun](#note_15):
> 
> > Monotonic time is always on. 
> 
> I'm not sure what this means.  Perhaps it means that we change the behavior of `reactor.callLater` to mean "call this after N seconds", rather than "call this at wallclock time now plus N seconds"?  Why would be do that, rather than introduce a second scheduling API?  Phrase your answer without contracting CompatibilityPolicy.

What effects would this have on application? The effect that I am concerned about is "what happens when the system clock gets adjusted?". If you could explain what other effects there could be (possibly related to the aforementioned ```DelayedCall.getTime```?), that would help me think about this.

As for the "what happens when the system clock gets adjusted?" issue, there are three possible classes of application:

• Those that rely on their events executing N seconds later and in the expected order, regardless of whether the system clock gets adjusted.
• Those that rely on the timing or ordering of their events getting changed when the system clock gets adjusted.
• Those that don't care when or in what order their events occur.

My intuition is that all or almost all applications fall into the first category, and changing the behavior of this API would be fixing a bug for them rather than introducing an incompatibility. For applications that fall into the second category, changing this would be incompatibly changing the API.

Apparently other people have a different intuition about this. Could you help me understand by providing examples of actual or plausible applications of the second category -- those that need their events to change timing or ordering because their system clock got reset?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

> I am unhappy with a new API because it means everyone has to change their existing code.

Not everybody.  Just applications that want the new behavior and not the old behavior - ie, applications that were broken before and want to be fixed.  Specifically, applications that were correct before require no change.

If possible, I would prefer to not have to fight about whether to preserve backwards compatibility or not.  That is why we have the CompatibilityPolicy, is it not?  If you think the change is "compatible enough" (whatever that might mean), then I think we need to find some way to resolve this kind of disagreement.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_15):

> A separate point which I don't see addressed anywhere on this ticket, yet, is how `CLOCK_MONOTONIC` interacts with the timeout parameter to `select`, `epoll_wait`, `kevent`, etc.  Perhaps this is obvious to everyone else and there's no special consideration required in the implementation; if so, it would be nice for someone to say so. :)

Some time ago, when investigating a related issue, I did a manual test of the behavior of `select`'s timeout on various platforms.  As I recall, it generally behaves as though it is using a monotonic timer; of course, these were very coarse-grained tests and I was just verifying them by eye.  I assume that some kind of run-as-root script that knew about both monotonic and wall clocks could be constructed to produce useful output.

Sometimes, according to rules I don't really understand, setting the system clock would instead cause a waiting `select` to time out immediately; a strict reading of [the relevant standard](http://pubs.opengroup.org/onlinepubs/000095399/functions/select.html) seems to say that it should not do this, but then, it says absolutely nothing about changes to the system clock.  However, I've not been able to reproduce this behavior in my more recent testing, and maybe it was a bug in some platform at the time (circa 3 years ago) which has since been fixed.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_20"></a>|@zooko commented|
|-|-|

Oh, I just noticed this:

http://docs.python.org/dev/library/time.html#time.monotonic

It looks like it will be in CPython 3.3 and uses the posix ```CLOCK_MONOTONIC_RAW``` if present, else the ```CLOCK_MONOTONIC```. On Windows it uses ```QueryPerformanceCounter```.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_21"></a>|@zooko commented|
|-|-|

This README from monoclock has an interesting TODO, to wit: https://github.com/ludios/Monoclock/blob/master/README.md

    Support buggy AMD chips, or expose a probablyBuggy() function that returns True if the monotonic clock is unreliable.

    Note: Chromium's base/time_win.cc just disables use of the monotonic clock on Athlon X2 CPUs with if (cpu.vendor_name() == "AuthenticAMD" && cpu.family() == 15

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_22"></a>|@zooko commented|
|-|-|

Here's a ctypes implementation: http://code.google.com/p/anacrolix/source/browse/projects/pimu/monotime.py

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_23"></a>|@zooko commented|
|-|-|

I told JP in person at Pycon about how I wrote my own workaround to protect my programs from system clock unsteadiness years ago. (I discovered the issue because the system clock was jumping around and causing my events to fire in a different order than I had told them to!) Here's the evidence of that: https://tahoe-lafs.org/trac/pyutil/browser/trunk/pyutil/increasing_timer.py?annotate=blame . The in-line comments mention it being 2001. I told him how when I switched from my own homegrown framework to Twisted I abandoned this measure because I trusted Twisted to (sooner or later) handle it for me.

I said that if there were an API named ```reactor.callAtDateAndTime```, that the programmer would expect the event to go off at the nominal date, on the *new* system clock, if the system clock changed after the call to ```reactor.callAtDateAndTime``` but before the event happened. On the other hand, since the API is ```reactor.callLater``` and the first argument is named ```delay```, the programmer expects the event to happen ```delay``` seconds afater the call to ```reactor.callLater```, regardless of whether the system clock changes in the interim.

JP said he didn't object to this change in principle, but he wanted someone to test it. Manually, that is, since it is awfully inconvenient to automatically, virtualizably, reproducibly test the behavior of your OS kernel...

To test this, I think you would write a Python script which starts two threads. One sleeps for a second while the other calls ```select``` with a given timeout, then the first wakes up and sets the system time (maybe by running ```sudo date --set``` in a subprocess). Then the human who is sitting there watching with his wristwatch in hand sees whether the duration of the call to ```select``` is affected by the change of clock. exarkun: is that what you meant?

There is an open issue about whether to use ```CLOCK_MONOTONIC``` or ```CLOCK_MONOTONIC_RAW``` on platforms that offer both (which means: on Linux). I don't mind that much either way. I think I would prefer ```CLOCK_MONOTONIC```, which if I understand correctly promises that even though it will never ever cause the clock to jump, either forward or backwards, it *will* cause the clock to run a bit faster or slower in the attempt to keep the clock's rate close to true. One hesitation I have about that is I don't know if some implementations of ```CLOCK_MONOTONIC``` might try to make the clock run faster in order to "catch up" to another clock, which to my mind violates the principle of the thing. On the other hand the drawback of ```CLOCK_MONOTONIC_RAW``` is that the clock will be chronically a little fast or a little slow (how much, I don't know. An experiment is probably in order), and may speed up when the clock is hotter and slow down when it is colder. That isn't ideal behavior, but at least is more likely that its behavior will be independent of bugs, overly helpful Linux kernel hackers, and potentially malicious ntp servers.

I'll bet there are very few people who know if the behavior of the "monotonic" clock on non-Linux platforms is more like linux's ```CLOCK_MONOTONIC``` or more like linux's ```CLOCK_MONOTONIC_RAW```...  I am not among them.

By the way the Boost and C++ folks have decided that the right name for this is a "steady clock" instead of a "monotonic clock". After all, the word "monotonic" means "it goes only in one direction", but leaves open the question in the mind of the programmer about whether it might jump *forward* in that direction in the attempt to catch up with someone else's standard of time. The word "steady" better conveys the truth: that it never jumps in either direction.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

[#5552](https://github.com/twisted/twisted/issues/5552) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

I had a look at [#5552](https://github.com/twisted/twisted/issues/5552) and the patch attached here and realized that neither of them dealt with OS X and one required an additional C extension.

I will attach a version that provides a monotonic clock for both Linux and OS X.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

Another thing to keep in mind as someone implements this is that system sleep will cause pretty much every platform's monotonic clock to leap forward; it doesn't increment while the system is in S3 or higher.  Unfortunately it looks like the suggested way to work around this is to simply look at the wall clock and trust that it's right!  So I started looking into what notifications various platforms can give you.

First up; linux.  It seems like time on linux is a real mess.  The kernel won't tell you anything about power management unless you're a driver.  Luckily, you can mostly ignore suspend on servers (at least for now) since they don't sleep.  Laptops, where this is more of a concern, are likely going to be running some kind of desktop environment, where you have access to the pleasant [Sleeping](http://upower.freedesktop.org/docs/UPower.html#UPower::Sleeping) and [Resuming](http://upower.freedesktop.org/docs/UPower.html#UPower::Resuming) D-Bus signals.  However, there's no particular obligation on linux to use UPower to manage your power, so suspend might happen through some other mechanism and you won't be told about it.  Too-clever-by-half users who like to cat things into /proc directly will break these notifications.  Right now there's no way to get told about `settimeofday`, but if we're lucky, [in the far-flung future you might be able to call sys_time_change_notify](http://article.gmane.org/gmane.linux.kernel/1061170/match=sys%5ftime%5fchange%5fnotify).  (FreeBSD seems to be in basically the same boat, without the hope for the future.)

Next: Windows.  As usual, everything is kind of gross, but possible.  The messages for [WM_POWERBROADCAST](http://msdn.microsoft.com/en-us/library/windows/desktop/aa373247(v=vs.85).aspx) and [WM_TIMECHANGE](http://msdn.microsoft.com/en-us/library/windows/desktop/ms725498(v=vs.85).aspx) should allow a process to figure out when power and time changes occur.  I'm not sure how much ctypes grossness this entails or how exactly this integrates with IOCP and win32reactor.

On OS X, everything happens through `NSNotificationCenter`, and assuming we're using `CFReactor`, I believe we can fairly easily receive [NSWorkspaceWillSleepNotification, NSWorkspaceDidWakeNotification](https://developer.apple.com/library/mac/#qa/qa1340/_index.html), and [NSSystemClockDidChangeNotification](https://developer.apple.com/library/mac/#documentation/Cocoa/Reference/Foundation/Classes/NSDate_Class/Reference/Reference.html#//apple_ref/occ/cl/NSDate) in python.

I don't think this information has ever been brought together in one place before, so hopefully this will serve as a useful reference to the future implementor of this feature for Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_27"></a>|@zooko commented|
|-|-|

If I call ```reactor.callLater(10, do_something)```, and one second later the laptop is suspended, and 24 hours later the laptop is unsuspended, do I expect for ```do_something``` to happen ASAP, or to wait 9 more seconds after unsuspend?

In other words, was the "10" to mean that 10 seconds have passed in the real world which the computer inhabits, or 10 seconds of awake computer time have passed?

I think maybe each alternative is sometimes intended by programmers, which means Twisted can't do a generically correct thing here unless we extend the API so the programmer can express what they want in this case.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph commented|
|-|-|

Replying to [zooko](#note_27):
> If I call ```reactor.callLater(10, do_something)```, and one second later the laptop is suspended, and 24 hours later the laptop is unsuspended, do I expect for ```do_something``` to happen ASAP, or to wait 9 more seconds after unsuspend?
> 
> In other words, was the "10" to mean that 10 seconds have passed in the real world which the computer inhabits, or 10 seconds of awake computer time have passed?
> 
> I think maybe each alternative is sometimes intended by programmers, which means Twisted can't do a generically correct thing here unless we extend the API so the programmer can express what they want in this case.

Quite so.

If you call `sleep(10); do_something();` kernel folks seem to have generally taken the opinion that that means "10 seconds of awake computer time".  There are some mailing list threads (which I can't find right now) where at least Linux kernel people have publicly said that this behavior is quite intentional, as if it were interpreted the other way, every laptop unsuspend would have every process waiting on a timeout (which is apparently a lot of them) getting scheduled to run instantly and consuming all your power.  Following that lead, I think that maybe `callLater` should mean "computer awake time" and we should add something higher-level like `callAtTime` which would take advantage of [comment:26 those power-management and clock-adjustment notifications I mentioned in my earlier comment].

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_29"></a>|@zooko commented|
|-|-|

+1

I read some of the linux kernel source in kernel/time/timekeeping.c, and I saw that it makes sure ```CLOCK_MONOTONIC``` and ```CLOCK_MONOTONIC_RAW``` show "computer awake time", but it offers a different function to get a steady measurement of time from the perspective of someone outside the computer.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_30"></a>|@itamarst commented|
|-|-|

As far as backwards compat, it seems like `DelayedCall.getTime` and `reactor.seconds` are the main issues, in case someone uses them in ways that interact with some other source of time (e.g. `time.time`). I would suggest returning a new object that:

  1. Includes a type info about what kind of time it is (clock vs. monotonic vs. fake for testing).
  1. When added/compared/etc. to floats or ints does a `DeprecationWarning`.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_31"></a>|@zooko commented|
|-|-|

Itamar: that sounds like a good idea to me!

**Comment by twisted-trac** [comment]
|<img alt="crmsystemet's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_32"></a>|crmsystemet commented|
|-|-|

I like this too Itamar. Will you be able to implement it ?

Replying to [itamar](#note_30):
> As far as backwards compat, it seems like `DelayedCall.getTime` and `reactor.seconds` are the main issues, in case someone uses them in ways that interact with some other source of time (e.g. `time.time`). I would suggest returning a new object that:
> 
>   1. Includes a type info about what kind of time it is (clock vs. monotonic vs. fake for testing).
>   1. When added/compared/etc. to floats or ints does a `DeprecationWarning`.
------------------------------------------------------
If you call sleep(10); do_something(); kernel folks seem to have generally taken the opinion that that means "10 seconds of awake computer time". There are some mailing list threads (which I can't find right now) where at least Linux kernel people have publicly said that this behavior is quite intentional, as if it were interpreted the other way, every laptop unsuspend would have every process waiting on a timeout (which is apparently a lot of them) getting scheduled to run instantly and consuming all your power. Following that lead, I think that maybe callLater should mean "computer awake time" [crm system](http://www.crmsystemet.dk) and we should add something higher-level like callAtTime  which would take advantage of those power-management and  clock-adjustment notifications I mentioned in my earlier comment.

**Comment by twisted-trac** [comment]
|<img alt="crmsystemet's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_33"></a>|crmsystemet commented|
|-|-|

If not i would consider helping out

Kind regards
Tom crm

Replying to [crm systemet](#note_32):
> I like this too Itamar. Will you be able to implement it ?
> 
> Replying to [itamar](#note_30):
> > As far as backwards compat, it seems like `DelayedCall.getTime` and `reactor.seconds` are the main issues, in case someone uses them in ways that interact with some other source of time (e.g. `time.time`). I would suggest returning a new object that:
> > 
> >   1. Includes a type info about what kind of time it is (clock vs. monotonic vs. fake for testing).
> >   1. When added/compared/etc. to floats or ints does a `DeprecationWarning`.
> ------------------------------------------------------
Replying to [crm systemet](http://www.crmsystemet.dk):
> If you call sleep(10); do_something(); kernel folks seem to have generally taken the opinion that that means "10 seconds of awake computer time". There are some mailing list threads (which I can't find right now) where at least Linux kernel people have publicly said that this behavior is quite intentional, as if it were interpreted the other way, every laptop unsuspend would have every process waiting on a timeout (which is apparently a lot of them) getting scheduled to run instantly and consuming all your power. Following that lead, I think that maybe callLater should mean "computer awake time"  and we should add something higher-level like callAtTime  which would take advantage of those power-management and  clock-adjustment notifications I mentioned in my earlier comment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun commented|
|-|-|

> Following that lead, I think that maybe callLater should mean "computer awake time"  crm system and we should add something higher-level like callAtTime which would take advantage of those power-management and clock-adjustment notifications I mentioned in my earlier comment.

This change breaks about half of the users of callLater.  Instead, please introduce a new API that means "call after N seconds of awake time".

Glyph is well aware of the CompatibilityPolicy so I'm not sure why he suggested this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_35"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_34):
> > Following that lead, I think that maybe callLater should mean "computer awake time"  crm system and we should add something higher-level like callAtTime which would take advantage of those power-management and clock-adjustment notifications I mentioned in my earlier comment.
> 
> This change breaks about half of the users of callLater.  Instead, please introduce a new API that means "call after N seconds of awake time".

Any users of `callLater` that are currently depending on this functionality are _already_ broken.

> Glyph is well aware of the CompatibilityPolicy so I'm not sure why he suggested this.

I think I'm going to retract my suggestion, and suggest fixing it the other way instead - by scheduling a wakeup somehow at unsleep.

The reason for fixing it this way instead is the expected behavior of `reactor.seconds()` (specifically, its mapping to real time).

But the reason why I suggested it in light of CompatibilityPolicy is that currently, if you have a reactor _today_ with only one timer scheduled with a `callLater(n, ...)` and nothing else going on, `n` will be interpreted by Twisted as a timeout, that timeout will be passed on to `select` or something, most operating systems we support will interpret that as 10 seconds of "computer awake time", and you won't get called.  We currently have no facility to wake up when the system wakes up; we'll just do it by accident if there's some socket activity or another, shorter timer timeout, and then all the other timers will accidentally fix themselves.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_36"></a>|@zooko commented|
|-|-|

Replying to [exarkun](#note_34):
> 
> This change breaks about half of the users of callLater.  Instead, please introduce a new API that means "call after N seconds of awake time".

"This change" here means that the countdown stops while the computer is asleep? Or does it mean that the countdown is unaffected by someone setting the system clock?

In either case, how can you tell that this change would break about half of the current users?

I don't know how to get insight into what behavior the current users are expecting/relying-on. I'd certainly like to! So please teach me.

The other ways to look at the compatibility issues are: 1. what the current documentation says it does, and 2. what the current behavior is.

First, the documentation. [It currently says](https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IReactorTime.html#callLater):

    Parameters	delay	the number of seconds to wait. (type: float)

To my ears, this means that the countdown is _not_ affected by resetting the system clock, and this is silent on the question of whether the countdown stops when the machine sleeps.

Second, the current behavior. According to [comment:28 comment:28], the current behavior in response to the computer sleeping, on Linux, is that the countdown stops while the computer is asleep. According to [comment:35 comment:35], the current behavior in response to the system clock getting changed, on Linux, depends on whether another event causes `select()` to return before the elapsed time, or not.

So I don't understand why exarkun says that the Compatibility Policy tells us to make the new version of `callLater` change its countdown in response to the system clock changing. As far as I understand, the current behavior of `callLater` in response to the system clock changing is unpredictable, and the documentation says that it will *not* change its countdown in response to the system clock changing, so it seems to me that Compatibility requires us to make the new version of `callLater` _not_ change its countdown in response to the system clock, and instead add a different API for events which are intended to happen at a certain system-clock-time instead of events which are intended to happen after a certain elapsed-time-from-now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_37"></a>|@glyph commented|
|-|-|

Hello again Zooko!

Replying to [zooko](#note_36):
> Replying to [exarkun](#note_34):
> > 

> I don't know how to get insight into what behavior the current users are expecting/relying-on. I'd certainly like to! So please teach me.

One major piece of behavior that users expect and rely on is that `reactor.seconds()` corresponds to wall-clock time, and that you can do math on that and call `datetime.fromtimestamp` on it and get a valid value.  Similarly, `IDelayedCall.getTime()` should return that same sort of value.

If we were to allow a "mode" where the reactor runs with a monotonic clock, we would break any code relying on this assumption because a monotonic clock will give values that correspond to a "local" time that cannot (almost by definition) be reliably mapped to a global calendar such as the Gregorian calendar.  We could change the _way_ that such code breaks in response to the system clock changing, but without a channel to communicate information about what the values returned by `seconds()` and `getTime()` _mean_ it's not possible to write correct code.

> The other ways to look at the compatibility issues are: 1. what the current documentation says it does, and 2. what the current behavior is.
> 
> First, the documentation. [It currently says](https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IReactorTime.html#callLater):
> 
>     Parameters	delay	the number of seconds to wait. (type: float)
> 
> To my ears, this means that the countdown is _not_ affected by resetting the system clock, and this is silent on the question of whether the countdown stops when the machine sleeps.

To mine, it just sounds like the documentation is ignoring all of these issues :).  The behavior is effectively undefined, because we have to transform the `delay` parameter to the `IDelayedCall.getTime()` value, which itself [is documented as](https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IDelayedCall.html#getTime)

> Returns	time in seconds since epoch (a float).

> Second, the current behavior. According to [comment:28 comment:28], the current behavior in response to the computer sleeping, on Linux, is that the countdown stops while the computer is asleep. According to [comment:35 comment:35], the current behavior in response to the system clock getting changed, on Linux, depends on whether another event causes `select()` to return before the elapsed time, or not.

The problem here is that in the face of a changing system clock, I don't think there is any possible way to make the various APIs all correspond to each other properly and to the documentation.

> So I don't understand why exarkun says that the Compatibility Policy tells us to make the new version of `callLater` change its countdown in response to the system clock changing. As far as I understand, the current behavior of `callLater` in response to the system clock changing is unpredictable, and the documentation says that it will *not* change its countdown in response to the system clock changing,

It just doesn't say what will happen in response to the system clock changing at all; the documentation assumes the system clock is correct at all times.

> so it seems to me that Compatibility requires us to make the new version of `callLater` _not_ change its countdown in response to the system clock, and instead add a different API for events which are intended to happen at a certain system-clock-time instead of events which are intended to happen after a certain elapsed-time-from-now.

I think that the only realistic way out of this situation is to introduce _two_ new APIs: `callAfter` and `callAt`.  `callAfter` would use the monotonic clock, always, and would return `IDelayedCallPlus` instances indicating that it uses a monotonic clock; `callAt` would use the wall clock, always, and would return `IDelayedCallPlus` instances indicating that its time represents wall clock time.

This would also resolve another problem with `callLater` that we haven't talked about: `callLater` always _must_ call `.seconds()` itself, making every timed call in Twisted require a `gettimeofday` or equivalent syscall.  These definitely add up in heavily time-sensitive applications, and I have seen them in profiles.  `callAt` would allow the application to decide when to call `.seconds()`, and although `callAfter` would need to get a monotonic timestamp, it could retrieve it using the (at least potentially) relatively cheaper `CLOCK_MONOTONIC_RAW` or `QueryPerformanceCounter` APIs.

With the two APIs clearly separated like this, it becomes clear what the behavior of each should be. Calls scheduled with `callAt` are invoked at the relevant wall clock time, so if the system clock changes, those change with it.  Calls scheduled with `callAfter` are invoked at their monotonic clock delta regardless of system clock changes.

In order to maintain a sane correspondence with expectations for `IDelayedCall.getTime` values from `callLater`, `callLater` in this regime would have to be implemented as a compatibility wrapper that would call `seconds()` and then call `callAt`, which would mean `callLater` calls _do_ move with the system clock; which is what I think exarkun was getting at in his comment 2 years ago.

Once the reactor is tracking both clocks, we can also _detect_ time shifts by looking for large deltas between the monotonic clock and the wall clock.  When we detect such a change, we can offer applications a callback to re-schedule or cancel their `callAt` callbacks, which would allow individual applications making heavy use of `callLater` able to cope with system clock changes without a total rewrite.

This still leaves the separate issue of _waking up at all_ when the system clock changes or the system wakes up from sleep, but I think we could fix each of those issues separately.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_38"></a>|@glyph commented|
|-|-|

To summarize my previous comment:

 1. This ticket's description, as (previously) written, cannot be satisfied. There's no way to run the reactor in a different mode, because a different mode would fundamentally change the meaning of the values returned and accepted by existing interfaces.  I think this is why we've stalled for so long about actually _doing_ anything about this.
 2. `callLater`'s semantics are tied to wall-clock time and so any compatible replacement (unfortunately) has to preserve that.  Therefore, long term, we might be looking at deprecating `callLater`.
 3. Since different applications have different relationships to time, it is not as simple as replacing `callLater` with `callAtTime`; we need _both_ `callAt` and `callAfter`.
 4. There is no _general_ solution to the question "what to do with a bunch of `callAt` callbacks if the system clock changes".  An application might want to _know_ if the system clock changes significantly, so we should have a callback for that.
 5. There are two related issues in the area of "the math on what to do with clock values might be right but we still have to figure out how to wake up".  On the mac these notifications are available via `NSNotificationCenter`, on Linux they're available via dbus, and I am not sure how they are made available on Windows:
   1. if the machine goes to sleep, `callAt` calls' time might have passed while it was asleep
   2. if the clock is adjusted, `callAt` calls' time might have passed during the period that was adjusted over.

I will update the summary and description next.
